runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/util/JMESPath.kt

```diff
@@ -56,3 +56,19 @@ public inline fun <reified T> Collection<Collection<T>>.flattenIfPossible(): Col
 @InternalApi
 public val <T> Collection<T>.length: Int
     get() = size
+
+/**
+ * Returns a JS type as a string
+ *
+ * See: [JMESPath spec](https://jmespath.org/specification.html#type)
+ */
+@InternalApi
+public fun Any?.type(): String = when (this) {
+    is String -> "string"
+    is Boolean -> "boolean"
+    is List<*>, is Array<*> -> "array"
+    is Short, is Int, is Long, is Float, is Double -> "number"
+    is HashMap<*, *>, is Any -> "object"
```

Comment:
question: why `HashMap` instead of the more-general `Map`?

Reply:
This is related to this [other function](https://github.com/awslabs/smithy-kotlin/blob/3934d48a9e084e23b1250501b0deb8dafe2a1c03/codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt#L506) that merges objects and returns a hash map. I didn't use a map for that, but I really don't know if it makes a difference. Is there a reason to use `Map` instead?

Reply:
I lied earlier, [HashMap](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-hash-map/) is an implementation of **Mutable**Map. I think it's better to use the base class as a type rather than an implementation of it. Let's see what others think about it / if it matters.

Reply:
If we're truly combining this logic with `is Any` then it doesn't matter—we should just remove `is HashMap<*, *>` since that type is a subtype of `Any`.

In principle though, @lauzadis is right: we should favor type checking at the highest reasonable level. Both in general and specifically for collections we should typically code to APIs (i.e., interfaces) rather than implementations (i.e., specific classes). I can see that `mergeProperties` creates a `HashMap` but there will be other kinds of maps in the object graph, including some created directly by deserializers which will delegate down to stdlib-internal implementations.